snapd-glib/snapd-notices-monitor.c

```diff
@@ -0,0 +1,166 @@
+                                   self->cancellable,
+                                   (GAsyncReadyCallback)monitor_cb,
+                                   self);
+}
+
+/**
+ * snapd_notices_monitor_start:
+ * @monitor: a #SnapdNoticesMonitor
+ * @cancellable: a #GCancellable
+ *
+ * Starts the asynchronous listening proccess, that will wait for new
+ * notices and emit a "notice-event" signal with the new notice as
+ * parameter.
+ *
+ * @return: TRUE if there was an error, FALSE if everything worked fine
+ * and the object is listening for events.
+ */
+gboolean snapd_notices_monitor_start(SnapdNoticesMonitor *self,
+                                 GCancellable *cancellable) {
+
+    g_return_val_if_fail(SNAPD_IS_NOTICES_MONITOR(self), TRUE);
+    self->cancellable = cancellable == NULL ? NULL : g_object_ref(cancellable);
+    begin_monitor(self);
+    return FALSE;
+}
+
+/**
+ * snapd_notices_monitor_get_client:
+ * @monitor: a #SnapdNoticesMonitor
+ *
+ * Returns the internal #SnapdClient used by this object. Useful if some
+ * kind of extra initialization must be done before begin to read notices.
+ * This client must NOT be used for anything else, because the get_notices
+ * call will block the connection until a response is generated. This client
+ * MUST be used only by this object.
+ *
+ * @return: (transfer none): a #SnapdClient
+ */
+SnapdClient *snapd_notices_monitor_get_client(SnapdNoticesMonitor *self) {
+    g_return_val_if_fail(SNAPD_IS_NOTICES_MONITOR(self), NULL);
+    return self->client;
+}
+
+static void snapd_notices_monitor_dispose(GObject *object) {
+    SnapdNoticesMonitor *self = SNAPD_NOTICES_MONITOR(object);
+
+    g_clear_object(&self->client);
+    g_clear_object(&self->cancellable);
+    g_clear_pointer(&self->last_date_time, g_date_time_unref);
+
+    G_OBJECT_CLASS(snapd_notices_monitor_parent_class)->dispose(object);
+}
+
+void snapd_notices_monitor_init(SnapdNoticesMonitor *self) {
+    self->client = snapd_client_new();
+}
+
+void snapd_notices_monitor_class_init(SnapdNoticesMonitorClass *klass) {
+    G_OBJECT_CLASS(klass)->dispose = snapd_notices_monitor_dispose;
+    g_signal_new ("notice-event",
+                  G_TYPE_FROM_CLASS (klass),
+                  G_SIGNAL_RUN_LAST,
+                  0,
+                  NULL,
+                  NULL,
+                  NULL,
+                  G_TYPE_NONE,
+                  1,
+                  SNAPD_TYPE_NOTICE);
+    g_signal_new ("error-event",
+                  G_TYPE_FROM_CLASS (klass),
+                  G_SIGNAL_RUN_LAST,
+                  0,
+                  NULL,
+                  NULL,
+                  NULL,
+                  G_TYPE_NONE,
+                  1,
+                  G_TYPE_ERROR);
+}
+
+SnapdNoticesMonitor *snapd_notices_monitor_new(void) {
+    SnapdNoticesMonitor *self = g_object_new(snapd_notices_monitor_get_type(), NULL);
+    self->last_date_time_seconds = -1.0;
+    return self;
+}
```

Comment:
No newline at end of file.

Reply:
Fixed. And added an extension to Code to make it automagic :-)